File: Yixuan/17_01_17/find_peak_element.cpp
class Solution {
public:
    /**
     * @param A: An integers array.
     * @return: return any of peek positions.
     * @author: Egbert Li
     * @link: http://www.lintcode.com/en/problem/find-peak-element/
     */
   int findPeak(vector<int> A) {
      int length = A.size();
      if (length == 0) {
          return 0;
      }
      int start = 0;
      int end = length - 1;
      while (start + 1 < end) {
         int mid = start + (end - start) / 2;
         if (A[mid] < A[mid - 1]) {
            end = mid;
         } else if (A[mid] < A[mid - 1]) {
            start = mid;
         } else {
            start = mid;
         }
      }
      if (A[start] > A[end]) {
         return start;
      }
      return end;
   }
};

File: Yixuan/17_01_17/find_peak_element.py
class Solution:
    #@param A: An integers list.
    #@return: return any of peek positions.
    #@link: http://www.lintcode.com/en/problem/find-peak-element/
    #@author: Egbert Li
    #$tag: Google
    def findPeak(self, A):
        if len(A) == 0:
            return 0
        while start + 1 < end:
            mid = int (start + (end - start) / 2)
            if A[mid] < A[mid - 1]:
                end = mid
            elif A[mid] < A[mid + 1]:
                start = mid
            else:
                start = mid
        if A[start] > A[end]:
            return start
        return end
